<a href="https://colab.research.google.com/github/freshanantaya-jae/rasberrypi/blob/main/html_table_059.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Get HTML Table to DataFrame
Data in tabular formats, such as CSV files, is very easy to process because the data relationships are straightforward. HTML tables are another format that allows us to easily analyze data. Copying or downloading HTML tables can be more difficult than you might think, as they often come with CSS styles that make reading the data difficult. Fortunately, the pandas library provides tools that allow us to easily extract data from HTML tables and place them in a DataFrame.

The following code example reads all HTML tables exist on the webpage at the specified URL as a list of dataframe.

In [3]:
import pandas as pd
import requests
from io import StringIO

url = "https://en.wikipedia.org/wiki/List_of_countries_and_dependencies_by_population_(United_Nations)"

# Add headers to mimic a web browser
headers = {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36"
}

# Fetch the HTML content using requests
try:
    response = requests.get(url, headers=headers)
    response.raise_for_status() # Raise an HTTPError for bad responses (4xx or 5xx)
    html_content = response.text

    # Read all tables from the webpage content
    # Wrap html_content in StringIO to suppress the FutureWarning
    tables = pd.read_html(StringIO(html_content))

    # tables is a list of DataFrames
    print(f"Number of tables found: {len(tables)}")

    # Get the first table
    df = tables[0]

    print(df.head())
except requests.exceptions.RequestException as e:
    print(f"Error fetching URL: {e}")
except Exception as e:
    print(f"An unexpected error occurred: {e}")

Number of tables found: 2
  Country or territory  Population (1 July 2022)  Population (1 July 2023)  \
0                World                8021407192                8091734930   
1                India                1425423212                1438069596   
2             China[a]                1425179569                1422584933   
3        United States                 341534046                 343477335   
4            Indonesia                 278830529                 281190067   

  Change (%) UN continental region[1] UN statistical subregion[1]  
0     +0.88%                        –                           –  
1     +0.89%                     Asia               Southern Asia  
2     −0.18%                     Asia                Eastern Asia  
3     +0.57%                 Americas            Northern America  
4     +0.85%                     Asia          South-eastern Asia  


Note: The code above might not work because modern web pages do not allow such a script to run. Use your preferred AI assistant to correct the code.

### Exercise 1
Use AI assistant to correct the code above. Replace the code above with the right one. Offline VS Code users might need to install dependencies (e.g., `requests`, `lxml`, or `beautifulsoup`).

## Data cleaning
The retrieved data usually needs to undergo a cleaning process. The data cleaning process often involves assigning appropriate header names. Long and difficult-to-spell header names can negatively impact coding efficiency.

### Exercise 2
Change column names to `Country`, `Jul2022`, `Jul2023`, `Change`, `Region`, `Subregion`.

In [4]:
# Exercise 2
# This block should print the dataframe with new column names
new_column_names = {
    'Country or territory': 'Country',
    'Population (1 July 2022)': 'Jul2022',
    'Population (1 July 2023)': 'Jul2023',
    'Change (%)': 'Change',
    'UN continental region[1]': 'Region',
    'UN statistical subregion[1]': 'Subregion'
}
df = df.rename(columns=new_column_names)
display(df.head())

,Country,Jul2022,Jul2023,Change,Region,Subregion
0,World,8021407192,8091734930,+0.88%,–,–
1,India,1425423212,1438069596,+0.89%,Asia,Southern Asia
2,China[a],1425179569,1422584933,−0.18%,Asia,Eastern Asia
3,United States,341534046,343477335,+0.57%,Americas,Northern America
4,Indonesia,278830529,281190067,+0.85%,Asia,South-eastern Asia


Often, tables contain unwanted characters such as references and footnotes. This is very common when scraping Wikipedia tables. You typically see things like:
- China[a]
- Tanzania[b]
- France[c]
We can apply `str.replace()` with `regex` to remove these unwanted characters.

### Exercise 3
Remove footnotes ([a], [b], [c], ...) and colonization ((United States), (France), (China), ...) from the 'Country' column. You may use AI assistant. After removing:
- *China[a]* should appear as *China*
- *Tanzania[b]* should appear as *Tanzania*
- ...
- *Hong Kong (China)* should appear as *Hong Kong*
- *Puerto Rico (United States)* should appear as *Puerto Rico*

In [5]:
# Exercise 3
# This block should clean the 'Country' column by removing footnotes and parenthetical information

# Remove text within square brackets (footnotes)
df['Country'] = df['Country'].str.replace(r'\[.*?\]', '', regex=True)

# Remove text within parentheses and the parentheses themselves (e.g., '(China)')
# This regex specifically targets ' (CountryName)' patterns that typically follow the main country name
df['Country'] = df['Country'].str.replace(r'\s*\(.*?\)', '', regex=True)

# Trim any leading/trailing whitespace that might be left after removal
df['Country'] = df['Country'].str.strip()

display(df.head())

,Country,Jul2022,Jul2023,Change,Region,Subregion
0,World,8021407192,8091734930,+0.88%,–,–
1,India,1425423212,1438069596,+0.89%,Asia,Southern Asia
2,China,1425179569,1422584933,−0.18%,Asia,Eastern Asia
3,United States,341534046,343477335,+0.57%,Americas,Northern America
4,Indonesia,278830529,281190067,+0.85%,Asia,South-eastern Asia


Important: The first important step before applying data cleaning is to observe the data and see if any cleaning can be applied. For example, using regex to remove all parentheses content may also remove other meaningful info.

### Exercise 4
Apply a filter so that the DataFrame displys only countries in South-eastern Asia and sort the data by country name. There should be 11 countries in total, starting with `Brunei` and ending with `Vietnam`. All columns must remain ('Country', 'Jul2022', 'Jul2023', 'Change', 'Region', 'Subregion'). You should assign filtered data to a new DataFrame (i.e., `df_sea`).

In [6]:
# Exercise 4
# This block should filter the dataframe for South-eastern Asia and sort by Country

df_sea = df[df['Subregion'] == 'South-eastern Asia'].sort_values(by='Country').reset_index(drop=True)

display(df_sea)

,Country,Jul2022,Jul2023,Change,Region,Subregion
0,Brunei,455370,458949,+0.79%,Asia,South-eastern Asia
1,Cambodia,17201724,17423880,+1.29%,Asia,South-eastern Asia
2,Indonesia,278830529,281190067,+0.85%,Asia,South-eastern Asia
3,Laos,7559007,7664993,+1.40%,Asia,South-eastern Asia
4,Malaysia,34695493,35126298,+1.24%,Asia,South-eastern Asia
5,Myanmar,53756787,54133798,+0.70%,Asia,South-eastern Asia
6,Philippines,113964338,114891199,+0.81%,Asia,South-eastern Asia
7,Singapore,5649885,5789090,+2.46%,Asia,South-eastern Asia
8,Thailand,71735329,71702435,−0.05%,Asia,South-eastern Asia
9,Timor-Leste,1369295,1384286,+1.09%,Asia,South-eastern Asia


## Data analysis
With pandas DataFrame, we can analyze the data with ease.

### Exercise 5
Perform a particular data analysis with the df dataframe (not df_sea). You may choose one of these options or define your own task.
- Find the total population in each continent in Jul 2023.
- Find the difference in population of each country between 2022 and 2023. You may add one more column to hold these numbers or modify data in the `Change` column.
- Change the population unit to 'million' (e.g., 1425423212 should be converted to 1425.4M). You may add two more columns to hold these numbers or modify data in the `Jul2022` and `Jul2023` columns.
- ...

In [7]:
# Exercise 5: Find the total population in each continent in Jul 2023

population_by_region_2023 = df.groupby('Region')['Jul2023'].sum().reset_index()
population_by_region_2023 = population_by_region_2023.sort_values(by='Jul2023', ascending=False)
display(population_by_region_2023)

,Region,Jul2023
5,–,8091734930
2,Asia,4777234490
0,Africa,1480770521
1,Americas,1041794259
3,Europe,745602874
4,Oceania,45562783
